### PR TITLE
Add Openresty as an webserver option

### PR DESCRIPTION
**Pull request by TafkaMax**
I have successfully created and made Openresty use Modsecurity. Because Openresty is based on Nginx it was doable.

There are some nuances.

----

Because the official Openresty Docker Image is created differently than the Nginx one, then there is no usage of 'nginx' user. Openresty does not use 'docker-entrypoint' logic, so I copied that logic from the original Nginx repository.

If you have any more questions I am happy to answer them.

I also added a simple gitlab-ci.yml file, because that is what I use with the integrated gitlab registry, but it might be too specific to add.

Currently I only added the Alpine build option and only for linux/amd64.

### CODE REVIEW

File: openresty/Dockerfile-alpine [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    LUA_JIT_VERSION="2.1-20230410"
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat as build
+
+ARG MODSEC_VERSION=3.0.9 \
+    OPENRESTY_VERSION \
+    NGINX_VERSION \
+    LUA_JIT_VERSION 
+
+# Remove pcre2 support for openresty, until it gets update to pcre2.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:15]: Is this sentence correct? Did you maybe mean to write "... updated to pcre3"?

File: openresty/Dockerfile-alpine [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        pcre-dev \
+        yajl-dev \
+        gd \
+        gd-dev \
+        zlib-dev
+
+
+WORKDIR /sources
+
+# Download modsecurity and compile it.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:51]:
```suggestion
# Download ModSecurity and compile it.
```

File: openresty/Dockerfile-alpine [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    cd ModSecurity; \
+    ARCH=$(gcc -print-multiarch); \
+    sed -ie "s/i386-linux-gnu/${ARCH}/g" build/ssdeep.m4; \
+    sed -ie "s/i386-linux-gnu/${ARCH}/g" build/pcre.m4; \
+    ./build.sh; \
+    ./configure --with-yajl --with-ssdeep --with-lmdb --with-geoip --enable-silent-rules --with-pcre=/usr/bin/pcre-config; \
+    make install; \
+    strip /usr/local/modsecurity/lib/lib*.so*
+
+# We use master Modsec version
```
Comment (theseion) [inline openresty/Dockerfile-alpine:63]:
```suggestion
# We use latest version of the ModSecurity nginx connector.
```

File: openresty/Dockerfile-alpine [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+# Instead of Nginx download the bundled openresty for this version of openresty base image. We can use the openresty already installed on the base image and get the compile options from there. For building the nginx module we can use the bundled nginx.
+# Specify the LUAJIT_LIB-s that are already installed in the base image.
+# use the includes from the bundled version. This is where LUA JIT VERSION is important. Because it differs from release to release (probably).
```
Comment (theseion) [inline openresty/Dockerfile-alpine:66]:
```suggestion
# Use the includes from the bundled version. This is where LUA_JIT_VERSION comes in to play because the version differs from release to release (probably).
```

File: openresty/Dockerfile-alpine [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    ARCH=$(gcc -print-multiarch); \
+    sed -ie "s/i386-linux-gnu/${ARCH}/g" build/ssdeep.m4; \
+    sed -ie "s/i386-linux-gnu/${ARCH}/g" build/pcre.m4; \
+    ./build.sh; \
+    ./configure --with-yajl --with-ssdeep --with-lmdb --with-geoip --enable-silent-rules --with-pcre=/usr/bin/pcre-config; \
+    make install; \
+    strip /usr/local/modsecurity/lib/lib*.so*
+
+# We use master Modsec version
+# Instead of Nginx download the bundled openresty for this version of openresty base image. We can use the openresty already installed on the base image and get the compile options from there. For building the nginx module we can use the bundled nginx.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:64]: Why is this necessary? Doesn't the Openresty image also include nginx?

```suggestion
# Download Openresty with bundled nginx. We can then get the compilation options for nginx from the installed Openresty and then build the ModSecurity nginx connector using the bundled nginx.
```

File: openresty/Dockerfile-alpine [hunk 6]
```diff
@@ -0,0 +1,253 @@
+# Current latest Openresty version is here.
+# Also specify the Nginx core used. (Openresty, adds their own version numbers after the nginx one, so we can't use the same number for both openresty and nginx)
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.2" \
+    NGINX_VERSION="1.21.4" \
+    LUA_JIT_VERSION="2.1-20230410"
```
Comment (theseion) [inline openresty/Dockerfile-alpine:6]: Declare the ModSecurity version at the top so you can access it in all stages.

```suggestion
    LUA_JIT_VERSION="2.1-20230410"
ARG MODSEC_VERSION=3.0.9
```

File: openresty/Dockerfile-alpine [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat as build
+
+ARG MODSEC_VERSION=3.0.9 \
```
Comment (theseion) [inline openresty/Dockerfile-alpine:10]: Use value of top-level ARG:

```suggestion
ARG MODSEC_VERSION
```

File: openresty/Dockerfile-alpine [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    curl -SL https://github.com/coreruleset/coreruleset/releases/download/v${CRS_RELEASE}/coreruleset-${CRS_RELEASE}.tar.gz.asc -o coreruleset-${CRS_RELEASE}.tar.gz.asc; \
+    gpg --fetch-key https://coreruleset.org/security.asc; \
+    gpg --verify coreruleset-${CRS_RELEASE}.tar.gz.asc v${CRS_RELEASE}.tar.gz; \
+    tar -zxf v${CRS_RELEASE}.tar.gz --strip-components=1 -C /opt/owasp-crs; \
+    rm -f v${CRS_RELEASE}.tar.gz coreruleset-${CRS_RELEASE}.tar.gz.asc; \
+    mv -v /opt/owasp-crs/crs-setup.conf.example /opt/owasp-crs/crs-setup.conf
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat
+
+ARG MODSEC_VERSION=3.0.9
```
Comment (theseion) [inline openresty/Dockerfile-alpine:119]: Use value from top-level ARG:
```suggestion
ARG MODSEC_VERSION
```

File: openresty/Dockerfile-alpine [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+LABEL maintainer="Felipe Zipitria <felipe.zipitria@owasp.org>"
```
Comment (theseion) [inline openresty/Dockerfile-alpine:121]: If you agree to maintain this image, then please change this label.

File: openresty/Dockerfile-alpine [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+COPY --from=build /usr/local/modsecurity/lib/libmodsecurity.so.${MODSEC_VERSION} /usr/local/modsecurity/lib/
+COPY --from=build /usr/local/openresty/nginx/modules/ngx_http_modsecurity_module.so /usr/local/openresty/nginx/modules/ngx_http_modsecurity_module.so
+COPY --from=build /usr/share/TLS/server.key /usr/local/openresty/nginx/conf/server.key
+COPY --from=build /usr/share/TLS/server.crt /usr/local/openresty/nginx/conf/server.crt
+COPY --from=build /usr/share/TLS/dhparam-* /etc/ssl/certs/
+COPY --from=build /etc/modsecurity.d/unicode.mapping /etc/modsecurity.d/unicode.mapping
+COPY --from=build /etc/modsecurity.d/modsecurity.conf /etc/modsecurity.d/modsecurity.conf
+COPY --from=crs_release /opt/owasp-crs /opt/owasp-crs
+# We use the templating mechanism from the nginx image here,
+# as set up by owasp/modsecurity-docker
```
Comment (theseion) [inline openresty/Dockerfile-alpine:197]:
```suggestion
```

File: openresty/Dockerfile-alpine [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+COPY --from=build /usr/local/modsecurity/lib/libmodsecurity.so.${MODSEC_VERSION} /usr/local/modsecurity/lib/
+COPY --from=build /usr/local/openresty/nginx/modules/ngx_http_modsecurity_module.so /usr/local/openresty/nginx/modules/ngx_http_modsecurity_module.so
+COPY --from=build /usr/share/TLS/server.key /usr/local/openresty/nginx/conf/server.key
+COPY --from=build /usr/share/TLS/server.crt /usr/local/openresty/nginx/conf/server.crt
+COPY --from=build /usr/share/TLS/dhparam-* /etc/ssl/certs/
+COPY --from=build /etc/modsecurity.d/unicode.mapping /etc/modsecurity.d/unicode.mapping
+COPY --from=build /etc/modsecurity.d/modsecurity.conf /etc/modsecurity.d/modsecurity.conf
+COPY --from=crs_release /opt/owasp-crs /opt/owasp-crs
+# We use the templating mechanism from the nginx image here,
```
Comment (theseion) [inline openresty/Dockerfile-alpine:196]: We no longer use owasp/modsecurity-docker. I'll create a PR to remove that from out other files.
```suggestion
# We use the templating mechanism from the nginx image here.
```

File: openresty/Dockerfile-alpine [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+# as set up by owasp/modsecurity-docker
+# Everything from templates except the nginx confiugration is the same for openresty.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:198]:
```suggestion
# Everything from templates, except the nginx configuration, is the same for Openresty.
```

File: openresty/Dockerfile-alpine [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+COPY nginx/templates/conf.d /usr/local/openresty/nginx/templates/conf.d
+COPY nginx/templates/includes /usr/local/openresty/nginx/templates/includes
+COPY nginx/templates/modsecurity.d /usr/local/openresty/nginx/templates/modsecurity.d
+# Copy the nginx configuration separately
+COPY openresty/templates/nginx.conf.template /usr/local/openresty/nginx/templates
+#
```
Comment (theseion) [inline openresty/Dockerfile-alpine:204]:
```suggestion
```

File: openresty/Dockerfile-alpine [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+# Copy the nginx configuration separately
+COPY openresty/templates/nginx.conf.template /usr/local/openresty/nginx/templates
+# 
+COPY src/etc/modsecurity.d/modsecurity-override.conf /usr/local/openresty/nginx/templates/modsecurity.d/modsecurity-override.conf.template
+COPY src/etc/modsecurity.d/setup.conf /usr/local/openresty/nginx/templates/modsecurity.d/setup.conf.template
+COPY src/bin/healthcheck /usr/local/bin/healthcheck
+COPY openresty/docker-entrypoint.d/*.sh /docker-entrypoint.d/
+COPY src/opt/modsecurity/activate-plugins.sh /docker-entrypoint.d/94-activate-plugins.sh
+COPY src/opt/modsecurity/activate-rules.sh /docker-entrypoint.d/95-activate-rules.sh
+COPY openresty/docker-entrypoint.sh /
```
Comment (theseion) [inline openresty/Dockerfile-alpine:211]: Please move all the Openresty stuff to a separate block after the other `COPY` statements. That will make it easier to maintain and merge changes.

File: openresty/Dockerfile-alpine [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        libxml2-dev \
+        lmdb-dev \
+        moreutils \
+        tzdata \
+        pcre \
+        # Alpine needs GNU 'sed' because the 'sed' program shipped with busybox does not support 'z' parameter for separating lines with a 'NUL' character.
+        sed \
+        yajl; \
+    # Install lua modules here
+    # Possible TODO make it a variable if you want to build this Dockerfile yourself.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:230]:
```suggestion
```

File: openresty/Dockerfile-alpine [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+    ln -s /usr/local/modsecurity/lib/libmodsecurity.so.${MODSEC_VERSION} /usr/local/modsecurity/lib/libmodsecurity.so.3; \
+    ln -s /usr/local/modsecurity/lib/libmodsecurity.so.${MODSEC_VERSION} /usr/local/modsecurity/lib/libmodsecurity.so; \
+    ln -sv /opt/owasp-crs /etc/modsecurity.d/; \
+    chmod -R g=u /var/log/ /var/run/ /usr/local/openresty/nginx/ /etc/modsecurity.d/
+
+EXPOSE 80
+
+HEALTHCHECK CMD /usr/local/bin/healthcheck
+
+# Openresty does not use NGINX dockerfile.
```
Comment (theseion) [inline openresty/Dockerfile-alpine:249]:
```suggestion
# The Openresty image is not based on the nginx image.
```

File: .gitlab-ci.yml
Comment (theseion) [inline .gitlab-ci.yml:1]: I'd rather not have this file in our repository. It's also pretty straight forward to adapt our github workflow to a gitlab style workflow.

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by theseion** [comment]
Thanks @TafkaMax. I have some comments but as it looks now, we could probably also build the Openresty image. We would update CRS and ModSecurity versions but would rely on you to update Openresty specific dependencies.

**Comment by fzipi** [comment]
@TafkaMax Can you address Max's comments?

**Comment by TafkaMax** [comment]
> Welcome back @TafkaMax ! Can you start addressing comments here, so we can merge this one?

Will do. I will plan do address the comments within a week.

File: openresty/Dockerfile-alpine (hunk 6 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine:6]: Should be OK now. Added MODSEC_VERSION at the start of the dockerfile and removed from other places.

File: openresty/Dockerfile-alpine (hunk 5 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine]: I have updated the comment.

I download the bundle mostly for the 'configure' script. I did not find this in the premade Openresty image under /usr/local/openresty. I guess they removed unneccessary build files from their image. As we need to add a new nginx module we need to download the version specific src of Openresty to do that.

**Comment by fzipi** [comment]
Can you resolve the conversation for the things addressed already? Thanks! Looking better :)

File: openresty/Dockerfile-alpine [hunk 17]
```diff
@@ -0,0 +1,254 @@
+# Current latest Openresty version is here.
+# Also specify the nginx core used. (Openresty, adds their own version numbers after the nginx one, so we can't use the same number for both Openresty and nginx)
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.3" \
```
Comment (theseion) [inline openresty/Dockerfile-alpine:4]:
```suggestion
ARG OPENRESTY_VERSION="1.25.3.1" \
```

File: openresty/Dockerfile-alpine [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+    NGINX_VERSION="1.21.4" \
```
Comment (theseion) [inline openresty/Dockerfile-alpine:5]:
```suggestion
    NGINX_VERSION="1.25.3" \
```

File: openresty/Dockerfile-alpine [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+    LUA_JIT_VERSION="2.1-20230410" \
```
Comment (theseion) [inline openresty/Dockerfile-alpine:6]:
```suggestion
    LUA_JIT_VERSION="2.1-20231117" \
```

File: openresty/Dockerfile-alpine [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+    MODSEC_VERSION="3.0.11"
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat as build
+
+ARG MODSEC_VERSION \
+    OPENRESTY_VERSION \
+    NGINX_VERSION \
+    LUA_JIT_VERSION 
+
+# Remove pcre2 compliation of ModSecurity for openresty, until openresty gets updated to support pcre2. Relevant github issue: https://github.com/openresty/lua-nginx-module/issues/1984
```
Comment (theseion) [inline openresty/Dockerfile-alpine:16]:
```suggestion
# Remove pcre2 compilation of ModSecurity for openresty, until openresty gets updated to support pcre2. Relevant github issue: https://github.com/openresty/lua-nginx-module/issues/1984
```

File: openresty/Dockerfile-alpine [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+    gpg --verify coreruleset-${CRS_RELEASE}.tar.gz.asc v${CRS_RELEASE}.tar.gz; \
+    tar -zxf v${CRS_RELEASE}.tar.gz --strip-components=1 -C /opt/owasp-crs; \
+    rm -f v${CRS_RELEASE}.tar.gz coreruleset-${CRS_RELEASE}.tar.gz.asc; \
+    mv -v /opt/owasp-crs/crs-setup.conf.example /opt/owasp-crs/crs-setup.conf
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat
+
+ARG MODSEC_VERSION
+
+LABEL maintainer="Taavi Ansper <taaviansperr@gmail.com>"
```
Comment (theseion) [inline openresty/Dockerfile-alpine:122]: Is this e-mail address correct? It appears to contain one more `r` than necessary.

File: openresty/templates/.gitkeep
Comment (theseion) [inline openresty/templates/.gitkeep:1]: The directory isn't empty, so you can delete this file. Or is there a reason for you to keep it?

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by theseion** [comment]
Sorry for the long wait @TafkaMax. I think you done a good job and if you are taking responsibility of the image then I don't see why we shouldn't offer it as part of the default build.

File: openresty/Dockerfile-alpine (hunk 21 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine:122]: Yes the email is correct.

File: openresty/templates/.gitkeep
Comment (TafkaMax) [reply openresty/templates/.gitkeep:1]: This can be deleted. I will do it.

File: openresty/Dockerfile-alpine (hunk 20 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine:16]: Because nginx was updated to 1.25 for openresty I will revisit the issue. The reason for openresty using 1.21 of nginx was because that was the last version to support pcre. The change to nginx 1.25 means that they have overcome that issue and I will need to change the compilation to use pcre2.

File: openresty/Dockerfile-alpine (hunk 17 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine:4]: Will change. Finally openresty has been updated to a newer version of nginx.

Comment (theseion) [reply openresty/Dockerfile-alpine:4]: Sorry, I didn't mean to put additional work on your plate, I simply wanted the versions to be consistent across all images.

Comment (TafkaMax) [reply openresty/Dockerfile-alpine:4]: No worries

File: openresty/Dockerfile-alpine (hunk 19 above)
Comment (TafkaMax) [reply openresty/Dockerfile-alpine]: Removed LUA_JIT_VERSION as this variable is not needed anymore

**Comment by TafkaMax** [comment]
Updated the image to version 1.25.3.

Now openresty can use pcre2 again, which is nice.

Tested in my environment and build successfully and works aswell.

File: README.md [hunk 22]
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
...
+⚠️ We changed tags to [support production usage](https://github.com/coreruleset/modsecurity-crs-docker/issues/67). Now, if you want to use the "rolling version", use the tag `owasp/modsecurity-crs:nginx-alpine` or `owasp/modsecurity-crs:apache-alpine` or `owasp/modsecurity-crs:openresty-alpine-fat`. If you need a stable long term image, use the one with the full CRS version, in addition to the build date in `YYYYMMDDHHMM` format, example `owasp/modsecurity-crs:3.3.5-nginx-alpine-202209141209` or `owasp/modsecurity-crs:3.3.5-apache-alpine-202209141209` or `owasp/modsecurity-crs:3.3.5-openresty-alpine-fat-202308051209` for example. You have been warned.
```
Comment (theseion) [inline README.md:31]:
```suggestion
⚠️ We changed tags to [support production usage](https://github.com/coreruleset/modsecurity-crs-docker/issues/67). Now, if you want to use the "rolling version", use the tag `owasp/modsecurity-crs:nginx-alpine` or `owasp/modsecurity-crs:apache-alpine` or `owasp/modsecurity-crs:openresty-alpine-fat`. If you need a stable long term image, use the one with the full CRS version, in addition to the build date in `YYYYMMDDHHMM` format (e.g., `owasp/modsecurity-crs:3.3.5-nginx-alpine-202209141209`).
```

File: README.md [hunk 23]
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
...
+
+### Notes regarding Openresty version of this image.
+
+The Openresty version of modsecurity-crs-docker currently only builds an image based on **Alpine Linux**. The Dockerfile for Openresty is located in the [docker-openresty github page](https://github.com/openresty/docker-openresty/blob/master/alpine/Dockerfile.fat).
```
Comment (theseion) [inline README.md:35]:
```suggestion
We currently only provide a version of the Openresty image based on **Alpine Linux**. The Dockerfile for Openresty resides in the [docker-openresty repository](https://github.com/openresty/docker-openresty/blob/master/alpine/Dockerfile.fat).
```

File: README.md [hunk 24]
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
...
 
 ## Supported architectures
 
-We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official apache and nginx ones, we can only support the architectures they support.
+We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official apache, nginx and openresty ones, we can only support the architectures they support.
```
Comment (theseion) [inline README.md:39]:
```suggestion
We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official Apache httpd, nginx and Openresty ones, we can only support the architectures they support.
```

File: README.md [hunk 25]
```diff
@@ -218,6 +229,10 @@ Note: Apache access and metric logs can be disabled by exporting the `nologging=
 | TIMEOUT  | Number of seconds for a keep-alive client connection to stay open on the server side (Default: `60s`) |
 | WORKER_CONNECTIONS  | Maximum number of simultaneous connections that can be opened by a worker process (Default: `1024`) |
 
+### Openresty ENV Variables
+
+Openresty uses the same ENV variables as the nginx version.
```
Comment (theseion) [inline README.md:234]:
```suggestion
Openresty uses the same environment variables as the nginx version.
```

File: openresty/Dockerfile-alpine [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+    PROXY_SSL_VERIFY=off \
+    PROXY_SSL_OCSP_STAPLING=off \
+    SERVER_NAME=localhost \
+    SSL_PORT=443 \
+    TIMEOUT=60s \
+    WORKER_CONNECTIONS=1024 \
+    # Change this from normal nginx setup. Do not add /usr/lib or /lib
+    LD_LIBRARY_PATH=/usr/local/lib:/usr/local/openresty \
+    NGINX_ENVSUBST_OUTPUT_DIR=/usr/local/openresty/nginx/conf \
+    # overridden variables
```
Comment (theseion) [inline openresty/Dockerfile-alpine:179]: Removed just recently.
```suggestion
```

File: openresty/Dockerfile-alpine [hunk 27]
```diff
@@ -0,0 +1,252 @@
...
+    USER=root \
```
Comment (theseion) [inline openresty/Dockerfile-alpine:180]: Removed just recently.
```suggestion
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
Awesome, thank you!
